src/templates_parser-cached_files.adb

```diff
@@ -339,12 +340,13 @@ package body Cached_Files is
          return;
       end if;
 
-      case Mode is
-         when Used =>
-            T.Info.Next.Used := T.Info.Next.Used + 1;
-         when Released =>
-            T.Info.Next.Used := T.Info.Next.Used - 1;
-      end case;
+      Ada.Assertions.Assert
+        (Check => Mode in Used | Released, Message => "Value outside expected value set");
+      if Mode in Used then
```

Comment:
Well, I'm not sure to like this! I understand that there is a gnatcheck rule for that but having a case statement with 2 alternatives at least ensure that when a new value is added the case statement will become invalid (compiler error) and so will require a review & change. This is far safer than the alternative you propose as the `else` part will catch all future use.

Reply:
I do agree with the doubts about the case to if statement rule.
The assert was added to partly address these doubts: the assert check is at run-time while the original case check was at compile time.

But the discussion of the quality of the rule is left as an internal discussion for AdaCore: which rules should be offered and why?
This proof of concept shows that besides automatically reporting issues (as gnatcheck and codepeer do), 
tools can also automatically repair the issues.
In particular, the find and replace can be written in a notation that Ada programmers will easily understand:
For example
```
Find_And_Replace (File_Path,
                        Analyze_Fragment ("not (not $S_Cond)", Expr_Rule).Root,
                        "$S_Cond");
```
```
Find_And_Replace (File_Path,
                        Analyze_Fragment ("not ($S_Var in $M_Values)", Expr_Rule).Root,
                        "$S_Var not in $M_Values");
```
```
Find_And_Replace (File_Path,
                        Analyze_Fragment
                          ( "if $S_Cond then"
                           & " return true;"
                           & "else"
                           & " return false;"
                           & "end if;",
                           If_Stmt_Rule).Root,
                        PRETTY_PRINT_ON
                        & "return $S_Cond;"
                        & ASCII.CR & ASCII.LF
                        & PRETTY_PRINT_OFF);
```
and
```
        Find_And_Replace (File_Path,
                          Analyze_Fragment
                            (  "case $S_Expr is "
                             & "when $M_Value_Set1 => $M_Stmts1;"
                             & "when $M_Value_Set2 => $M_Stmts2;"
                             & "end case;",
                             Case_Stmt_Rule).Root,
                          PRETTY_PRINT_ON
                          & "Ada.Assertions.Assert (Check => ($S_Expr) in $M_Value_Set1 | $M_Value_Set2,"
                          & "                       Message => ""Value outside expected value set"");" & ASCII.CR & ASCII.LF
                          -- Ada doesn't support multiline string literals, so we 
                          -- can't include the string representation of the expected value sets
                          & "if ($S_Expr) in $M_Value_Set1 then" & ASCII.CR & ASCII.LF
                          & "   $M_Stmts1;"                      & ASCII.CR & ASCII.LF
                          & "else"                               & ASCII.CR & ASCII.LF
                          & "   $M_Stmts2;"                      & ASCII.CR & ASCII.LF
                          & "end if;"
                          & ASCII.CR & ASCII.LF
                          & PRETTY_PRINT_OFF);
```